### ISSUE TITLE
Function 'strlen' not found when both strlen functions available

### ISSUE THREAD

**Issue by bladewolf55** [peachpiecompiler/peachpie#165]
I recently worked on a personal project to convert blog posts from my old Simple PHP Blog into Markdown. A challenge would be producing the HTML from the custom-marked-up posts. I knew there was a PHP function in the site code to do exactly that, and rather than rewrite the function in C# I wondered if I could just use it. That led me to Peachpie and the peach.exe utility. However, getting it to work may have uncovered a bug, or at least an unexpected error. So, here's all the information, which I hope helps.

My goal was to convert one file, `sb_formatting.php`, to use as a dll. 
The original version would compile with minor revision, once I included Library.Graphics in peach.exe. But, I couldn't use the resulting dll because no functions were contained in a class.

I surrounded all functions in a class, i.e.

```php
<?php 
class SPHPB {
    //code here
}
?>
```
which yielded these errors:

```txt
Call to undefined function: 'replace_url_tag'
Call to undefined function: 'replace_img_tag'
Call to undefined function: 'replace_html_tag'
```
I'm not a PHP developer, so based on research I tried making all the functions public, and calling the error-producing functions with object notation: `$this->replace_url_tag`

This allowed the dll to compile, but running it using this code

```csharp
List<BlogEntry> entries = new List<BlogEntry>();
var files = System.IO.Directory.GetFiles(contentFolder, "entry*.txt*", SearchOption.AllDirectories);
entries.AddRange(files.Select(a => Convert(File.ReadAllText(a))));
var ctx = Pchp.Core.Context.CreateEmpty();
SPHPB sphp = new SPHPB(ctx);
foreach (var entry in entries)
{
    entry.Html = sphp.blog_to_html(PhpValue.Create(entry.Content), PhpValue.False, PhpValue.False).ToString(ctx);
}
var x = entries;
```
I got this error
`System.ArgumentException: 'Function 'strlen' not found!'`

on the noted line

```php
// Search for the beginning part of the tag.
$str_offset = strpos( $str, $tag_begin );
while ( $str_offset !== false ) {
	// Store sub_string before the tag.
	$str_out = $str_out . substr( $str, 0, $str_offset );
	// Store sub_string after the tag.
	$str = substr( $str, $str_offset + strlen( $tag_begin ) );
//ERROR ON THE ABOVE LINE
```
Quite a bit of experimenting later, I found that if I commented out one of the two strlen functions in Peachpie.Library.Strings.cs, I could run my app without error as expected.

```csharp
/// <summary>
/// Returns the length of a string.
/// </summary>
public static int strlen(string x) => (x != null) ? x.Length : 0;

/// <summary>
/// Returns the length of a string.
/// </summary>
//public static int strlen(PhpString x) => x.Length;
//COMMENTED OUT ABOVE
```
Attached files:

* **sb_formatting_orig.php.txt**  This is the slightly modified original script.
* **sb_formatting_compiles.php.txt** The version I eventually used.
* **entry091111-082951.txt** One of my blog posts.

[sb_formatting_orig.php.txt](https://github.com/peachpiecompiler/peachpie/files/1687947/sb_formatting_orig.php.txt)

[sb_formatting_compiles.php.txt](https://github.com/peachpiecompiler/peachpie/files/1687950/sb_formatting_compiles.php.txt)

[entry091111-082951.txt](https://github.com/peachpiecompiler/peachpie/files/1687960/entry091111-082951.txt)

Simple PHPBlog Source: <https://sourceforge.net/projects/sphpblog/>

**Comment by jakubmisek** [comment]
Good use case!

You are missing one annoying API call (that will be hopefully not needed in a future version of API):
```csharp
Pchp.Core.Context.AddScriptReference(Assembly.Load(new AssemblyName("sphp"))); // this reads referenced functions, classes and constants required by your compiled sphp library, including definition of strlen()
```

The reason is
1. we need two `strlen` functions (for Unicode and for single-byte strings).
2. Since it is questionable which one should be called during compile-time, compiler postpones the resolution to run-time.
3. Runtime handles the resolution and could not find the declaration in these magic references.

**Comment by jakubmisek** [comment]
Second,
```
Call to undefined function: 'replace_url_tag'
Call to undefined function: 'replace_img_tag'
Call to undefined function: 'replace_html_tag'
```
are not errors but warnings. These functions are simply not known during compile time, they are probably defined in another PHP file.

**Comment by bladewolf55** [comment]
Thanks for the information. I uncommented the strlen function, recompiled peach.exe as well as my dll, updated my program with the API call, and it's working as expected. Yay!

Regarding the undefined functions, those functions are in the sb_formatting.php file, so definitely available. To get them working, I had to use object notation, e.g. $this->replace_url_tag. I assumed that was because of converting the file into a class. (I'm not a PHP programmer.)

**Comment by jakubmisek** [comment]
good.

Yes, when you put functions into class, you have to access them using object notation. However I suggest to keep them as global functions (outside the class), there is no need to put them inside the class.

**Comment by bladewolf55** [comment]
Understood. I did it that way, too, with only the blog_to_html function inside the class. 

Nice tool. I'll try to blog about my experience soon. I won't get nearly as many readers as Scott Hanselman, though. :-)

**Comment by bfistein** [comment]
Hi @bladewolf55 that would be great! Definitely drop us a link when you do, we're always happy to share those.
